### PR TITLE
Refactoring multiprocessing indexation

### PR DESCRIPTION
**Pull request by nikitagromov**
Refactor multiprocess indexation, avoid issues with parallel application start inside consumer process
Related PR https://github.com/geniusproject/nefertari-sqla/pull/59

### CODE REVIEW

File: nefertari/scripts/es.py [hunk 1]
```diff
@@ -72,15 +74,18 @@ def main(argv=sys.argv):
     else:
         model_names = split_strip(options.models)
 
-    queue = get_processing_queue(model_names, processes, options)
-    pool = Pool(processes)
-    processors = map(lambda i: ESProcessor(options=options, queue=queue, pid=i),
+    BaseObject.metadata.bind.dispose()
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: Does this dispose of the current connection? Why do you do this?

File: nefertari/scripts/es.py [hunk 2]
```diff
@@ -72,15 +74,18 @@ def main(argv=sys.argv):
...
+
+    consumers = map(lambda i: TaskConsumer(options=options, manager=manager),
                      range(0, processes))
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: Wouldn't this work?
```python
consumers = [TaskConsumer(options=options, manager=manager) for i in range(processes)]
```

File: nefertari/scripts/es.py [hunk 3]
```diff
@@ -107,6 +112,19 @@ def _check_results(result):
         assert len(initial_dict[model_name]) == items_count
 
 
+def process_tasks(consumers, producer, manager):
+    consumers = list(consumers)
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: Is this necessary?

File: nefertari/scripts/es.py [hunk 4]
```diff
@@ -149,84 +169,134 @@ def recreate_index(registry):
...
+        while not self.tasks.empty():
+            time.sleep(1)
+
+        self.tasks.close(consumers=range(0, self.consumers_count))
+
+    def wait_for_processes(self):
+        self.barrier.wait()
+
+    def process_finished(self):
+        self.barrier.wait()
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: These two methods do the same thing

**Review by nmenardg — CHANGES_REQUESTED** [review]

File: nefertari/scripts/es.py (hunk 1 above)
Comment (nikitagromov) [reply nefertari/scripts/es.py]: Yes, we need to close all existed connections for avoid issue with concurrent access to one conection. http://docs.sqlalchemy.org/en/latest/faq/connections.html#how-do-i-use-engines-connections-sessions-with-python-multiprocessing-or-os-fork

Comment (nmenardg) [reply nefertari/scripts/es.py]: Ok, can you please comment in the code?

**Review by nmenardg — APPROVED** [review]

**Pull request merged**